pkg/lifecycle/service.go

```diff
@@ -173,18 +241,26 @@ func (s *Service) stopForceful(ctx context.Context, pl *Instance) error {
 	return nil
 }
 
-// StopAll will ask all the pipelines to stop gracefully
+// StopAll will ask all the running pipelines to stop gracefully
 // (i.e. that existing messages get processed but not new messages get produced).
 func (s *Service) StopAll(ctx context.Context, reason error) {
-	for _, pl := range s.instances {
-		if pl.GetStatus() != StatusRunning && pl.GetStatus() != StatusRecovering {
+	s.m.Lock()
+	instances := s.runningPipelines
+	s.m.Unlock()
```

Comment:
This only copies the pointer to the map, so the map could still change while we iterate over it. We'll need to hold the lock while we iterate over it (an alternative would be to make a copy of the map, but I don't think that's needed):
```suggestion
	s.m.Lock()
	defer s.m.Unlock()
```